proofs.go

```diff
@@ -119,37 +126,154 @@ func (proof *SparseCompactMerkleProof) Unmarshal(bz []byte) error {
+		return fmt.Errorf("invalid closest proof: %w", err)
+	}
+	return nil
+}
 
-	return true
+// SparseCompactMerkleClosestProof is a compressed representation of the SparseMerkleClosestProof
+type SparseCompactMerkleClosestProof struct {
+	Path             []byte                    // the path provided to the ProveClosest method
+	FlippedBits      [][]byte                  // the index of the bits flipped in the path during tree traversal
+	Depth            []byte                    // the depth of the tree when tree traversal stopped
+	ClosestPath      []byte                    // the path of the leaf closest to the path provided
+	ClosestValueHash []byte                    // the value hash of the leaf closest to the path provided
+	ClosestProof     *SparseCompactMerkleProof // the proof of the leaf closest to the path provided
+}
+
+func (proof *SparseCompactMerkleClosestProof) validateBasic(spec *TreeSpec) error {
+	// Do a basic sanity check on the proof on the fields of the proof specific to
+	// the compact proof only.
+	//
+	// When the proof is de-compacted and verified, the sanity check for the
+	// de-compacted proof should be executed.
+
+	// ensure no compressed fields are larger than the path size
+	maxSliceLen := minBytes(spec.ph.PathSize() * 8)
+	if len(proof.Depth) > maxSliceLen {
+		return fmt.Errorf("invalid depth: %d", proof.Depth)
+	}
+	for _, i := range proof.FlippedBits {
+		if len(i) > maxSliceLen {
+			return fmt.Errorf("invalid flipped bit index: %d", bytesToInt(i))
+		}
+	}
+	// perform a sanity check on the closest proof
+	if err := proof.ClosestProof.validateBasic(spec); err != nil {
+		return fmt.Errorf("invalid closest proof: %w", err)
+	}
+	return nil
+}
+
+// Marshal serialises the SparseCompactMerkleClosestProof to bytes
+func (proof *SparseCompactMerkleClosestProof) Marshal() ([]byte, error) {
+	buf := bytes.NewBuffer(nil)
+	enc := gob.NewEncoder(buf)
+	if err := enc.Encode(proof); err != nil {
+		return nil, err
+	}
+	return buf.Bytes(), nil
+}
+
+// Unmarshal deserialises the SparseCompactMerkleClosestProof from bytes
+func (proof *SparseCompactMerkleClosestProof) Unmarshal(bz []byte) error {
+	buf := bytes.NewBuffer(bz)
+	dec := gob.NewDecoder(buf)
+	return dec.Decode(proof)
 }
 
 // VerifyProof verifies a Merkle proof.
-func VerifyProof(proof *SparseMerkleProof, root, key, value []byte, spec *TreeSpec) bool {
-	result, _ := verifyProofWithUpdates(proof, root, key, value, spec)
-	return result
+func VerifyProof(proof *SparseMerkleProof, root, key, value []byte, spec *TreeSpec) (bool, error) {
+	result, _, err := verifyProofWithUpdates(proof, root, key, value, spec)
+	return result, err
 }
 
 // VerifySumProof verifies a Merkle proof for a sum tree.
-func VerifySumProof(proof *SparseMerkleProof, root, key, value []byte, sum uint64, spec *TreeSpec) bool {
+func VerifySumProof(proof *SparseMerkleProof, root, key, value []byte, sum uint64, spec *TreeSpec) (bool, error) {
```

Comment:
You should generate a godoc for this in the future! All these exported functions will show up.